copyoptions.go

```diff
@@ -0,0 +1,86 @@
+/*
+Copyright The ORAS Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package oras
+
+import (
+	"context"
+	"math"
+
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"golang.org/x/sync/semaphore"
+)
+
+const defaultConcurrencyLimit = 3
+
+type copyOptions struct {
+	limiter             *semaphore.Weighted
+	copyCallbackHandler func(ctx context.Context, desc ocispec.Descriptor) error
```

Comment:
abount naming: this is more like a `copyDoneHandler`.

Reply:
Also can we add another handler called we a node being found existed during copying (see [here](https://github.com/oras-project/oras-go/blob/main/copy.go#L86-L88))?

It will be a compliment to this copy call back function, client can handle 
- copied desc through `copyDoneHandler`
- skipped desc through `copySkippedHandler`


Feel free to resolve directly if it's out of this PR's scope.

Reply:
What would be the E2E UX for oras CLI? Do we want something like below:
```
$ oras cp $src $dst 
sha256:abc123: Completed
sha256:def456: Already Exists
```

Reply:
Yes, and the `copySkippedHandler` can be used for the `Already Exists` output

Reply:
How about other scenarios like blob not found error?

Reply:
We can have a design meeting for this scenario.

Reply:
Adding a E2E UX for the design discussion when using oras pull command to download a manifest config:

Suppose the target config don't have a title name stored in the annotation or the stored title doesn't match the expected download file name, the oras cli client need to 1) rewrite the title annonation 2) call oras-go for copying 3) print out the config name and digest

`copyDoneHandler` won't satify the need since it only covers step 1), we should also provide another handler and `beforeCopyHandler` and maybe as a step further, pack those two in one copy option property.